compilers: various environment fixes

matz-e opened the pull request:
Add common environment variables to all module generating configs. This
should set MPI_C* for all modules, even user-generated ones, since the
application stage configuration is propagated to the end user.

Following the official Spack documentation, we should actually populate
the `modules` key of the `compilers.yaml`. For this, remove all empty
lines, and append fresh ones to matching specs (account for LLVM
changing to clang).

See also https://spack.readthedocs.io/en/latest/getting_started.html#compilers-requiring-modules

matz-e committed:
compilers: various environment fixes

Add common environment variables to all module generating configs. This
should set MPI_C* for all modules, even user-generated ones, since the
application stage configuration is propagated to the end user.

Following the official Spack documentation, we should actually populate
the `modules` key of the `compilers.yaml`. For this, remove all empty
lines, and append fresh ones to matching specs (account for LLVM
changing to clang).

See also https://spack.readthedocs.io/en/latest/getting_started.html#compilers-requiring-modules

matz-e committed:
sed: delimiters…

matz-e committed:
sed: snafu, my bad

matz-e committed:
d-bug

matz-e committed:
more d-bug

matz-e committed:
more d-bug

matz-e committed:
moar d-bug

matz-e committed:
provider-index: only consider concrete specs

matz-e committed:
capture more.

matz-e committed:
d-bug: moar info

pramodk commented:
> Add common environment variables to all module generating configs. This
> should set MPI_C* for all modules, even user-generated ones, since the
> application stage configuration is propagated to the end user.

I read the description but still need to understand why we need to duplicate those variables in every stage.

matz-e commented:
> > Add common environment variables to all module generating configs. This
> > should set MPI_C* for all modules, even user-generated ones, since the
> > application stage configuration is propagated to the end user.
> 
> I read the description but still need to understand why we need to duplicate those variables in every stage.

Ideally, these would just end up in a higher-level configuration that covers all stages. Easier to have it everywhere than to forget that these are needed later.

matz-e committed:
compilers: some tuning.

matz-e committed:
providers: just skip uncomparable stuff

matz-e committed:
don't occupy a full butler-node

pramodk commented on deploy/Jenkinsfile:
```diff
@@ -83,7 +83,7 @@ def run_stages() {
 pipeline {
     agent {
-        label 'bb5-full'
+        label 'bb5'
```
no longer full node only?

matz-e replied:
We compile on the cluster nodes, so we don't need a full node any longer.

pramodk commented on deploy/deploy.lib:
```diff
@@ -352,6 +352,18 @@ configure_compilers() {
+        full_module_path="$(${cmd}|&sed -ne '2{s/:$// p}')"
+        spec=${spec%%[+ ]*}             # truncate variants
+        spec=${spec/llvm/.clang}        # LLVM identifies as clang with some garbage in front
+        spec=${spec/20.0.2/19.1.2.254}  # Intel identifies as an older version
```
I don't understand details here but noticed version hardcoding here...

matz-e replied:
Yes… normally, Intel would just append a build number or so to the version, but for `intel@20`, `icc` actually quotes a version number starting with `19`!

pramodk reviewed (dismissed):
LGTM

matz-e committed:
another approach?

pramodk commented on var/spack/repos/builtin/packages/nvhpc/package.py:
```diff
@@ -96,9 +96,6 @@ def setup_run_environment(self, env):
                                   self.version, 'compilers'))
 
-        # pgi/nvc doesn't support -isystem flag
-        env.set('DUMB_COMPILER', 'absolutely')
-
```
thanks! :)

matz-e committed:
provider index: smarter filtering?

matz-e committed:
deploy: assorted fixes

Remove the additional quote that may appear in front of clang, add some
documentation bits everywhere.

matz-e committed:
fix: forgot a (

matz-e committed:
deploy: no more quote in front of clang :)

matz-e committed:
deploy: explicitly modify environment for %intel

matz-e committed:
F GNU/Stallman's BASH

The abominations just keep on giving…

matz-e committed:
fix api

matz-e committed:
Add blah to py
